Add support for manufacturer-specific clusters, commands, and attributes

triller-telekom opened the pull request:
Replacement for PR #498

Closes #429

Also-by: Henning Sudbrock <henning.sudbrock@telekom.de>
Signed-off-by: Stefan Triller <stefan.triller@telekom.de>

triller-telekom commented:
@cdjackson FYI: Here is the commit which adjusts Hennings changes to the current ZSS-1.2.0 version. I have already tested it with a Centralite 3310-D device which provides humidity and temperature values. Please have a look.

cdjackson commented:
This pull request **introduces 2 alerts** when merging c202c6de1709330bf6c06f02514068ca61d1bcc5 into 339eef60cbc6b3b959ef985685bd61ce23542da8 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-d95b7dbdf315e09c3d39ad0d91bc30da6d7788fb)

**new alerts:**

* 2 for Spurious Javadoc @param tags

---

*Comment posted by [LGTM.com](https://lgtm.com)*

cdjackson commented:
Thanks @triller-telekom . I had a quick look and it looks similar to what I also did in this branch (https://github.com/zsmartsystems/com.zsmartsystems.zigbee/commit/5f223190a6bed78250f4c8980edea8a5f1598fa4) as I mentioned in my email to Michael last week ;)

The main difference I think is that I don't like the new version of sendDefaultResponse. Given that most commands are not manufacturer specific, I don't think we should require all uses of this to include ```null``` on the end of the call. Maybe you can take a look at this? This should avoid changes to the OTA cluster as well.

I think if we can align on that, then we can merge.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/zsmartsystems/com.zsmartsystems.zigbee?pullRequest=606) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/zsmartsystems/com.zsmartsystems.zigbee?pullRequest=606) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: hsudbrock<br/>:x: triller-telekom<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/zsmartsystems/com.zsmartsystems.zigbee?pullRequest=606) it.</sub>

triller-telekom commented:
I have updated the PR to be rebased on your newly force-pushed dev-1.2.0 branch and I added one commit which reverts the changes to `sendDefaultResponse`.

Basically I extended the existing `sendDefaultResponse` method with the manufacturer specific parameter and made the original method without that parameter call the new one with `null`.

@cdjackson Please have a look at this PR again.

cdjackson commented:
Sorry @triller-telekom can you rebase this - it's showing too many changes now as I rebased master this morning. If you can rebase on Monday, I'll review without any further changes so we can get this merged.

triller-telekom commented:
@cdjackson I have rebased the commits again. Please have a look.

cdjackson commented:
This pull request **introduces 1 alert** when merging 7972c25c773bc8e5b9b714470e23e38708d8ae07 into f054cf9ea0f924e47768d21dc322b01d0ed1f880 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-dfdc61604c367a15c06637232fa028ca011dcd60)

**new alerts:**

* 1 for Spurious Javadoc @param tags

---

*Comment posted by [LGTM.com](https://lgtm.com)*

triller-telekom commented:
Thank you for your review. I have added a commit to address your comments. Feels free to squash the commits before merging. I just left them separate to simplify your reviewing efforts.

cdjackson commented on com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zcl/ZclCluster.java:
```diff
@@ -1285,4 +1413,50 @@ protected Object readSync(final ZclAttribute attribute) {
+    /**
+     * Adds additional attributes to the cluster (like, e.g., manufacturer-specific attributes).
+     */
+    public void addAttributes(Set<ZclAttribute> attributes) {
```
I think you missed this one?

cdjackson commented on com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zcl/ZclCluster.java:
```diff
@@ -1285,4 +1413,50 @@ protected Object readSync(final ZclAttribute attribute) {
+
+    /**
+     * Adds additional client commands to the cluster (like, e.g., manufacturer-specific commands).
+     */
```
I think you missed this one?

cdjackson commented:
Sorry @triller-telekom github was showing me outdated review data - having sent the comment, I noticed it was kindly telling me it was outdated, where for the others it showed the updated info. Strange...

Anyway - looks good - thanks.

triller-telekom commented:
Thanks for merging. No worries regarding github. I guess its because your comment was on the method declaration whereas my necessary changes where above that in the comments so in a different line. That's probably why github was not able to match them.
